src/main/java/eu/dissco/backend/service/SpecimenService.java

```diff
@@ -36,92 +41,91 @@ public class SpecimenService {
+    var linksNode = new JsonApiLinksFull(pageNumber, pageSize, hasNext, path);
+    var dataNode = hasNext ? dataNodePlusOne.subList(0, pageSize) : dataNodePlusOne;
+    return new JsonApiListResponseWrapper(dataNode, linksNode);
   }
 
-  public DigitalSpecimen getSpecimenById(String id) {
-    return repository.getLatestSpecimenById(id);
+  public JsonApiWrapper getSpecimenById(String id, String path) {
+    var dataNode = repository.getLatestSpecimenById(id);
+    return new JsonApiWrapper(dataNode, new JsonApiLinks(path));
   }
 
-  public List<DigitalSpecimen> search(String query, int pageNumber, int pageSize)
+  public JsonApiListResponseWrapper search(String query, int pageNumber, int pageSize, String path)
       throws IOException {
-    return elasticRepository.search(query, pageNumber, pageSize);
-  }
+    var specimensPlusOne = elasticRepository.search(query, pageNumber, pageSize+1);
+    boolean hasNext = specimensPlusOne.size() > pageSize;
+    var specimens = hasNext ? specimensPlusOne.subList(0, pageSize) : specimensPlusOne;
 
-  public List<AnnotationResponse> getAnnotations(String id) {
-    return annotationService.getAnnotationForTarget(id);
+    var linksNode = new JsonApiLinksFull(pageNumber, pageSize, hasNext, path);
+    return new JsonApiListResponseWrapper(specimens, linksNode);
   }
 
-  public DigitalSpecimen getSpecimenByVersion(String id, int version)
-      throws JsonProcessingException, NotFoundException {
-    var result = mongoRepository.getByVersion(id, version, "digital_specimen_provenance");
-    return mapResultToSpecimen(result);
+  public JsonApiListResponseWrapper getAnnotations(String id, String path) {
+    return annotationService.getAnnotationForTarget(id, path);
   }
 
-  private DigitalSpecimen mapResultToSpecimen(JsonNode result) {
-    var digitalSpecimen = result.get("digitalSpecimen");
-    var attributes = digitalSpecimen.get("ods:attributes");
-    return new DigitalSpecimen(
-        result.get("id").asText(),
-        result.get("midsLevel").asInt(),
-        result.get("version").asInt(),
-        Instant.ofEpochSecond(result.get("created").asInt()),
-        digitalSpecimen.get("ods:type").asText(),
-        digitalSpecimen.get("ods:physicalSpecimenId").asText(),
-        attributes.get("ods:physicalSpecimenIdType").asText(),
-        attributes.get("ods:specimenName").asText(),
-        attributes.get(ORGANISATION_ID).asText(),
-        attributes.get("ods:datasetId").asText(),
-        attributes.get("ods:physicalSpecimenCollection").asText(),
-        attributes.get("ods:sourceSystemId").asText(),
-        attributes,
-        digitalSpecimen.get("ods:originalAttributes"),
-        digitalSpecimen.get("ods:attributes").get("dwca:id").asText()
-    );
+  public JsonApiWrapper getSpecimenByVersion(String id, int version, String path)
+      throws JsonProcessingException, NotFoundException {
+    var specimen = mongoRepository.getByVersion(id, version, "digital_specimen_provenance");
```

Comment:
We manipulate the request from MongoDB. We cannot just through the whole JSON from mongo to the user. We will create different objects for elastic, db and mongo. We need to harmonise this data in the domain object and then use this in the JSON:API response

Reply:
Restored